### PR TITLE
Ports Goonchat

### PR DESCRIPTION
**Pull request by ghost**
Ports the following PRs from tg: https://github.com/tgstation/tgstation/pull/26965, https://github.com/tgstation/tgstation/pull/28354, https://github.com/tgstation/tgstation/pull/28670, https://github.com/tgstation/tgstation/pull/28750, https://github.com/tgstation/tgstation/pull/29056, https://github.com/tgstation/tgstation/pull/29336, https://github.com/tgstation/tgstation/pull/29667, and https://github.com/tgstation/tgstation/pull/29690

WOOO GOONCHAT

:cl:  
experimental: Ported goonchat from /tg/
/:cl:

FOR CODERS: DO NOT USE ```\icon[thing]``` anymore, use ```[icon2html(thing, person_or_list_of_people_seeing_this_icon)]```

### CODE REVIEW

**Comment by X-TheDark** [comment]
You are a madman.

**Comment by matskuman5** [comment]
What is Goonchat and why is it better than Oldchat

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
it's an uglier version of oldchat which can change font size

**Comment by ghost** [comment]
@matskuman5 see https://github.com/tgstation/tgstation/pull/26965

**Comment by ghost** [comment]
screenshot 1

**Comment by Partheo** [comment]
For anyone who's unaware, this allows RMB support in the text window.

**Comment by Xantam-13** [comment]
Some men just want to see the world burn.

File: code/modules/client/client_procs.dm [hunk 1]
```diff
@@ -345,6 +356,7 @@ var/next_external_rsc = 0
 	if(!tooltips)
 		tooltips = new /datum/tooltip(src)
 
+	winset(src, null, "reset=true")
```
Comment (X-TheDark) [inline code/modules/client/client_procs.dm]: What does this line do?

During testing, this causes the New Player panel thing where you pick Join Game, etc to disappear and you become unable to join a game in progress, even if you Reconnect. Commenting out this line made the New Player panel work again, but as I've said, I have no idea if removing this breaks something.

Comment (X-TheDark) [inline code/modules/client/client_procs.dm]: What does this line do?

During testing, this causes the New Player panel thing where you pick Join Game, etc to disappear and you become unable to join a game in progress, even if you Reconnect. Commenting out this line made the New Player panel work again, but as I've said, I have no idea if removing this breaks something.

Tested using the latest stable BYOND build (Release 511.1385)

File: code/modules/goonchat/browserOutput.dm [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+
+	for(var/message in messageQueue)
+		to_chat(owner, message)
+
+	messageQueue = null
+	sendClientData()
+
+	//do not convert to to_chat()
+	owner << "<span class=\"userdanger\">If you can see this, update byond.</span>"
```
Comment (X-TheDark) [inline code/modules/goonchat/browserOutput.dm]: This is not necessarily true. If players use the File->Reconnect option, it breaks the Goonchat. You definitely should add "Try the "Fix chat" verb in OOC tab." somewhere in this.

**Comment by Partheo** [comment]
Please don't make it a pref.
Overall this is a change for the better and the only reason why people want a pref is because they aren't used to the change and prefer the old way. Adding prefs shouldn't be the way we go about this.

**Comment by ghost** [comment]
@Partheo why? it's on by default, and it'll let people who prefer oldchat to use oldchat

**Comment by Partheo** [comment]
Give them a week and they'll all be used to it

**Comment by Alexandria404** [comment]
The system makes it impossible to post messages using ` or ´ among other buttons.
The messages simply don´t get posted.

By using ' instead I can get most messages to work but still some simply don´t get posted (tried in asay).
Furthermore players reported the following bugs:

"there are a bunch of bugs with the new chat
you can't check your own wounds currently
also it randomly scrolls up"

Furthermore I prefer oldchat a lot more. This style no matter of the font size is a lot more spaced on my laptop. I have a standard full HD screen.
It makes it immensly more difficult to follow the chat for me.

If at all possible make it a preference. There is no reason not to have the option if it is possible @partheo.

**Comment by ghost** [comment]
@Alexandria404 THERE IS A PREFERENCE... Game preferences > Chat Window.

click on it to toggle between oldchat and goonchat

**Comment by Alexandria404** [comment]
Found it. Thanks for the guidance.

**Comment by Alexandria404** [comment]
I did some testing.
People who have goonchat enabled simply don´t see messages that are posted with certain keys in them.
People who have oldchat enabled see them.

**Comment by ghost** [comment]
Closed in favor of #2939

**Pull request closed without merging**